### PR TITLE
#860: Migrate bin/improver- python to lib/improver/cli

### PR DESCRIPTION
**Pull request by benfitzpatrick**
Addresses #860.

Testing:
 - [ ] Ran tests and they passed OK

### CODE REVIEW

**Comment by arh89** [comment]
How come so many things changed after Codacy looked at the code? It this because Codacy wasn't looking at these files when they were in bin (under the old names)?

File: lib/improver/argparser.py [hunk 1]
```diff
@@ -137,6 +139,11 @@ def __init__(self, central_arguments=DEFAULT_CENTRALIZED_ARG_NAMES,
         central_arguments = [ArgParser.CENTRALIZED_ARGUMENTS[arg_name] for
                              arg_name in central_arguments]
 
+        if 'prog' not in kwargs:
+            improver_oper = pathlib.PurePosixPath(
+                sys.argv[0]).stem.replace("_", "-")
+            kwargs['prog'] = ("improver " + improver_oper)
```
Comment (arh89) [inline lib/improver/argparser.py]: ...I might be being dense, but why do we need this kwarg?

**Comment by arh89** [comment]
Since we're touching all the CLI files anyway, I think it would be nice if we could change the `main` definition to accept `argv` (i.e. `main(argv=None)`), and then pass `argv` to `parse_args`. This would keep the existing behaviour (deriving from `sys.argv` if `None` is passed in), but also allow us to pass in the command line args to the `main` function.

...This is still not as nice as having a properly defined interface, and doesn't separate out the IO from the processing, but it at least means we won't need to call the CLIs using subprocess or something equally as messy.
(Used in my prototypes, and the one by @abooton/@carolinejjones)

**Comment by arh89** [comment]
I notice there are some failures in the CLI tests (probably not related to the changes in this PR...):
```
Test Summary Report
-------------------
improver/tests/improver-ensemble-calibration/04_save_mean_and_var.bats                                 (Wstat: 256 Tests: 1 Failed: 1)
  Failed test:  1
  Non-zero exit status: 1
improver/tests/improver-ensemble-calibration/03_trunc_gaussian.bats                                    (Wstat: 256 Tests: 1 Failed: 1)
  Failed test:  1
  Non-zero exit status: 1
improver/tests/improver-ensemble-calibration/02_gaussian.bats                                          (Wstat: 256 Tests: 1 Failed: 1)
  Failed test:  1
  Non-zero exit status: 1
improver/tests/improver-ensemble-calibration/07_current_forecast_percentiles.bats                      (Wstat: 256 Tests: 1 Failed: 1)
  Failed test:  1
  Non-zero exit status: 1
improver/tests/improver-ensemble-calibration/05_current_forecast_probabilities.bats                    (Wstat: 256 Tests: 1 Failed: 1)
  Failed test:  1
  Non-zero exit status: 1
improver/tests/improver-feels-like-temp/00-null.bats                                                   (Wstat: 256 Tests: 1 Failed: 1)
  Failed test:  1
  Non-zero exit status: 1
improver/tests/improver-neighbour-finding/02-nearest-uk.bats                                           (Wstat: 256 Tests: 1 Failed: 1)
  Failed test:  1
  Non-zero exit status: 1
improver/tests/improver-neighbour-finding/03-nearest-land-constraint-uk.bats                           (Wstat: 256 Tests: 1 Failed: 1)
  Failed test:  1
  Non-zero exit status: 1
improver/tests/improver-neighbour-finding/05-nearest-land-constraint-minimum-dz-uk.bats                (Wstat: 256 Tests: 1 Failed: 1)
  Failed test:  1
  Non-zero exit status: 1
improver/tests/improver-neighbour-finding/04-nearest-minimum-dz-uk.bats                                (Wstat: 256 Tests: 1 Failed: 1)
  Failed test:  1
  Non-zero exit status: 1
improver/tests/improver-neighbour-finding/07-nearest-global.bats                                       (Wstat: 256 Tests: 1 Failed: 1)
  Failed test:  1
  Non-zero exit status: 1
improver/tests/improver-neighbour-finding/06-all-methods-uk.bats                                       (Wstat: 256 Tests: 1 Failed: 1)
  Failed test:  1
  Non-zero exit status: 1
improver/tests/improver-neighbour-finding/08-nearest-land-constraint-global.bats                       (Wstat: 256 Tests: 1 Failed: 1)
  Failed test:  1
  Non-zero exit status: 1
improver/tests/improver-neighbour-finding/10-nearest-land-constraint-minimum-dz-global.bats            (Wstat: 256 Tests: 1 Failed: 1)
  Failed test:  1
  Non-zero exit status: 1
improver/tests/improver-neighbour-finding/14-different-grid-metadata.bats                              (Wstat: 256 Tests: 1 Failed: 1)
  Failed test:  1
  Non-zero exit status: 1
improver/tests/improver-neighbour-finding/15-invalid-global-site.bats                                  (Wstat: 256 Tests: 1 Failed: 1)
  Failed test:  1
  Non-zero exit status: 1
improver/tests/improver-neighbour-finding/16-circular-coord-beyond-bounds.bats                         (Wstat: 256 Tests: 1 Failed: 1)
  Failed test:  1
  Non-zero exit status: 1
improver/tests/improver-neighbour-finding/09-nearest-minimum-dz-global.bats                            (Wstat: 256 Tests: 1 Failed: 1)
  Failed test:  1
  Non-zero exit status: 1
improver/tests/improver-neighbour-finding/11-all-methods-global.bats                                   (Wstat: 256 Tests: 1 Failed: 1)
  Failed test:  1
  Non-zero exit status: 1
improver/tests/improver-orographic-enhancement/00-null.bats                                            (Wstat: 256 Tests: 1 Failed: 1)
  Failed test:  1
  Non-zero exit status: 1
improver/tests/improver-percentiles-to-realizations/00-null.bats                                       (Wstat: 256 Tests: 1 Failed: 1)
  Failed test:  1
  Non-zero exit status: 1
improver/tests/improver-percentiles-to-realizations/05-reordering_wrong_options.bats                   (Wstat: 256 Tests: 1 Failed: 1)
  Failed test:  1
  Non-zero exit status: 1
improver/tests/improver-percentiles-to-realizations/06-rebading_wrong_option1.bats                     (Wstat: 256 Tests: 1 Failed: 1)
  Failed test:  1
  Non-zero exit status: 1
improver/tests/improver-percentiles-to-realizations/07-rebading_wrong_option2.bats                     (Wstat: 256 Tests: 1 Failed: 1)
  Failed test:  1
  Non-zero exit status: 1
improver/tests/improver-temporal-interpolate/00-null.bats                                              (Wstat: 256 Tests: 1 Failed: 1)
  Failed test:  1
  Non-zero exit status: 1
improver/tests/improver-time-lagged-ensembles/00-null.bats                                             (Wstat: 256 Tests: 1 Failed: 1)
  Failed test:  1
  Non-zero exit status: 1
improver/tests/improver-update-grid-metadata/00-null.bats                                              (Wstat: 256 Tests: 1 Failed: 1)
  Failed test:  1
  Non-zero exit status: 1
improver/tests/improver-uv-index/00-null.bats                                                          (Wstat: 256 Tests: 1 Failed: 1)
  Failed test:  1
  Non-zero exit status: 1
improver/tests/improver-weighted-blending/04-bothoptions_fail.bats                                     (Wstat: 256 Tests: 1 Failed: 1)
  Failed test:  1
  Non-zero exit status: 1
improver/tests/improver-weighted-blending/07-invalid_args_lin_nonlin.bats                              (Wstat: 256 Tests: 1 Failed: 1)
  Failed test:  1
  Non-zero exit status: 1
improver/tests/improver-weighted-blending/08-invalid_args_nonlin_lin.bats                              (Wstat: 256 Tests: 1 Failed: 1)
  Failed test:  1
  Non-zero exit status: 1
```
I notice conflicts with ensemble calibration, probably due to PR #859 being merged - but that doesn't explain the others...

**Comment by benfitzpatrick** [comment]
> How come so many things changed after Codacy looked at the code? It this because Codacy wasn't looking at these files when they were in bin (under the old names)?

Yep!

**Pull request merged**